website/auth.py

```diff
@@ -340,25 +338,27 @@ def destroy(user):
     @app.route('/auth/change_password', methods=['POST'])
     @requires_login
     def change_password(user):
-
         body = request.json
-        if not isinstance(body, dict):
-            return 'body must be an object', 400
-        if not isinstance(body.get('old_password'), str):
-            return 'body.old_password must be a string', 400
-        if not isinstance(body.get( 'new_password'), str):
-            return 'body.new_password must be a string', 400
 
-        if len(body['new_password']) < 6:
-            return 'password must be at least six characters long', 400
+        if not isinstance(body, dict):
+            return g.auth_texts.get('ajax_error'), 400
+        if not isinstance(body.get('old_password'), str) or not isinstance(body.get('password'), str):
+            return g.auth_texts.get('password_invalid'), 400
+        if not isinstance(body.get( 'password_repeat'), str):
+            return g.auth_texts.get('repeat_match_password'), 400
+        if len(body['password']) < 6:
+            return g.auth_texts.get('password_six'), 400
+        if body['password'] != body['password_repeat']:
+            return g.auth_texts.get('repeat_match_password'), 400
 
         # The user object we got from 'requires_login' doesn't have the password, so look that up in the database
         user = DATABASE.user_by_username(user['username'])
 
         if not check_password(body['old_password'], user['password']):
-            return 'invalid username/password', 403
+            print("Hier zou de error gegooid moeten worden")
+            return g.auth.texts.get('password_invalid'), 403
```

Comment:
@TiBiBa I found it.

```suggestion
            return g.auth_texts.get('password_invalid'), 403
```

Here's what I did to find this issue:

## Run a regular Hedy app in one terminal tab

```
$ python app.py
```

Note that it runs on port 8080.

## Run the failing test by itself in another terminal tab

```
$ env ENDPOINT=http://localhost:8080 pytest -q tests_e2e.py -k test_invalid_change_password
```

I mostly copy/pasted this line from `build-tools/github/validate-e2e`, filled in the `$PORT` with `8080`, and added `-k test_invalid_change_password` to *only* run the specific test that was failing.

## Notice the error in the app error output

Because this makes sure that the app logging output is only relevant to the test at hand (and not all other 30+ successful tests), it's now much easier to see the following error in the scrollback:

```
INFO:werkzeug:127.0.0.1 - - [16/Jan/2022 11:19:39] "POST /auth/change_password HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/rix0rrr/Dev/hedy/.env/lib/python3.8/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)

  (...bunch of irrelevant Flask innards...)

  File "/Users/rix0rrr/Dev/hedy/website/auth.py", line 125, in inner
    return f(current_user(), *args, **kws)
  File "/Users/rix0rrr/Dev/hedy/website/auth.py", line 360, in change_password
    return g.auth.texts.get('password_invalid'), 403
  File "/Users/rix0rrr/Dev/hedy/.env/lib/python3.8/site-packages/werkzeug/local.py", line 347, in __getattr__
    return getattr(self._get_current_object(), name)
AttributeError: '_AppCtxGlobals' object has no attribute 'auth'
```

Showing me that the error happens on:

```py
    return g.auth.texts.get('password_invalid'), 403   # AttributeError: '_AppCtxGlobals' object has no attribute 'auth'
```

And then looking at the corresponding code I can see it's a typo that needs to be `g.auth_texts`.

Reply:
Wow good catch! Thanks!